packages/components/src/ui/create-styles/create-style-system/utils.js

```diff
@@ -0,0 +1,186 @@
+ *
+ * @param {string} key The variable (key).
+ * @return {string} The token (CSS variable).
+ */
+export function createToken( key ) {
+	return `${ NAMESPACE }-${ kebabCase( key ) }`;
+}
+
+/**
+ * The primary mechanism to retrieve Style system configs values - values that
+ * have been transformed into CSS variables with a dedicated namespace.
+ *
+ * @example
+ * ```js
+ * get('colorAdmin'); // var(--wp-g2-color-admin, 'blue');
+ * ```
+ * @template {Record<string, string | number>} TConfig
+ * @template {Record<string, string | number>} TDarkConfig
+ * @template {Record<string, string | number>} THCConfig
+ * @template {Record<string, string | number>} TDarkHCConfig
+ * @template {string} TGeneratedTokens
+ * @param {keyof (TConfig & TDarkConfig & THCConfig & TDarkHCConfig) | TGeneratedTokens} key The config variable to retrieve.
+ * @return {string} The compiled CSS variable associated with the config key.
+ */
+export function get( key ) {
+	return `var(${ createToken( key.toString() ) })`;
+}
+
+/** @typedef {Record<string, string | number>} StyleConfigValues */
+/** @typedef {Record<string, string>} StyleConfig */
+
+/**
+ * Transforms a series of config values into set of namespaced CSS
+ * references for the Style system.
+ *
+ * @param {StyleConfigValues} values Style config values to transform into CSS variables.
+ * @return {StyleConfig} The set of CSS variables, transformed from config values.
+ */
+export function transformValuesToReferences( values = {} ) {
+	/** @type {StyleConfig} */
+	const next = {};
+	for ( const [ key, value ] of Object.entries( values ) ) {
+		const ref = `var(${ createToken( key ) }, ${ value })`;
+		next[ key ] = ref;
+	}
+	return next;
+}
+
+/**
+ * Transforms a series of config values into set of namespaced CSS
+ * variables for the Style system. These values can then be safely and predictable
+ * retrieved using the get() function.
+ *
+ * @param {StyleConfigValues} values Style config values to transform into CSS variables.
+ * @return {StyleConfig} The set of CSS variables, transformed from config values.
+ */
+export function transformValuesToVariables( values = {} ) {
+	/** @type {StyleConfig} */
+	const next = {};
+
+	for ( const [ key, value ] of Object.entries( values ) ) {
+		const ref = value;
+		next[ `${ createToken( key ) }` ] = ref?.toString();
+	}
+
+	return next;
+}
+
+/**
+ * Transforms a series of config values into set of namespaced CSS
+ * references for the Style system. These values are then transformed into
+ * a CSS style value (`string`) that can be injected into the DOM, within a
+ * <style> tag.
+ *
+ * @param {string} [selector=':root'] The selector to attach the config values to.
+ * @param {StyleConfigValues} values Style config values to transform into CSS variables.
+ * @param {boolean} isGlobal
+ * @return {string} Compiled innerHTML styles to be injected into a <style /> tag.
+ */
+export function transformValuesToVariablesString(
```

Comment:
This sound like things used by the ThemeProvider to generate the CSS variables, since we don't have that anymore, maybe there are some things to clean/remove here?

Reply:
Nope, this is used by `generateTheme` to apply the generated theme variables to the `:root` scope at the very least as well as the different "modes" offered by the new style system.

Reply:
so basically, this is just a duplication of what we currently do in this mixin https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_mixins.scss#L441 

I think we should just add the G2 variables there for now and avoid all the extra JS logic that runs at init.

Reply:
But aren't we trying to do CSS-in-JS? It seems counter intuitive that we would keep our variables in SASS in that case?

Reply:
There's a few things here:

- I think the logic of these function is just too complex for what it is and right now, why not just inject a static `<style>` element? 
- I'd like use to avoid having two ways of injecting these "theme styles", so we should consolidate in one way
- Right now the variables that constitute a "theme" are used not just in the "@wordpress/components" package, they can be used in all the other packages and these packages are not CSS in JS.

Reply:
In other words, If we really want to keep it in JS (which I don't feel strongly against), I think it should be:

 - its own package that all the other packages that use the variables should depend on. (Right now the dependency towards the SASS mixin is in the build script)
 - Simplify it's logic to just print the style properly.
 - Remove the SASS mixin. 

I feel it's a lot of work probably, more than just moving the G2 variables to the mixin for the moment.

Reply:
I think the last one is the most compelling reason to me.

I think it'd still be good to have the `ui.get` function to make variables easy to retrieve (and have the variables documented in JS and strictly typed so that people aren't accidentally introducing new variables or using non-existent ones)...

I'll have to think about how to do this but I agree we can remove this variable generation stuff 🤔 I just wish the original author were around to give their expert opinion on this 😞

Reply:
> Simplify it's logic to just print the style properly.

By this do you mean that the new package would just create the theme variables into the `:root` scope?

I'm not sure which is less work. Refactoring how the theme is created is an unknown variable to me.

Reply:
yes, basically do what `_default-custom-properties.scss` does now.

Reply:
Where would the new package's function to create the theme variables be called?

Reply:
I think my ideal solution for now, is just to remove all the theme generation, grab the CSS variables injected right now and move them to the existing mixing and call it a day. (assuming we work on a follow-up on removing the need for most of these variables and replace them with static values like discussed).

Reply:
I'd love other thoughts on this as well. This is something that can have some impact long term (WP user profiles, theming components) so yeah more thoughts and opinions here would be good. cc @mtias @jasmussen @mcsf @gziolo

Reply:
The thing is, there are a TON of variables that are generated. I'm not sure it's possible to just plop them into a mixin as some of them are runtime generated using `tinycolor`: 
https://github.com/ItsJonQ/g2/blob/97ef23ebe56a4968701a36ce6697f75803700134/packages/styles/src/theme/utils.js#L45

https://github.com/ItsJonQ/g2/blob/97ef23ebe56a4968701a36ce6697f75803700134/packages/styles/src/theme/tokens/colors.js#L196

https://github.com/ItsJonQ/g2/blob/97ef23ebe56a4968701a36ce6697f75803700134/packages/styles/src/theme/tokens/colors.js#L83

Reply:
@sarayourfriend would it make sense in this case to actually switch the priorities of the integration:

 1- First refactor the `ui.get` to avoid returning variables (and just return static values potentially from a theme context) instead of returning CSS variables (the main color can be kept as a variable)
 2- Do this removal later.


How much work would **1** here be. It feels to me that's important regardless because it prevents us from introducing these public variables inadvertently.

Reply:
So `ui.get` actually does just return the CSS custom property. Get is a one line function here: 

https://github.com/WordPress/gutenberg/blob/da56a34c0697e8f7dae3139ee7952e597f9e63de/packages/components/src/ui/create-styles/create-style-system/create-style-system.js#L138-L142

```ts
function get(
  key: keyof TConfig | keyof TDarkConfig | keyof THCConfig | keyof TDarkHCConfig | TGeneratedTokens
) {
  return `var(${ createToken( key.toString() ) })`;
}
```

As you can see it just calls `createToken` on the passed in `key` and generates the `var(...)`.

So is what you're proposing to change `get` to actually return the raw value of the variable rather than the CSS Custom Property? Just want to make sure I'm understanding.

I think this could be relatively simple... the theme is a flat object so I think `get` would just be turned into this:

```ts
function get(
  key: keyof TConfig | keyof TDarkConfig | keyof THCConfig | keyof TDarkHCConfig | TGeneratedTokens
) {
  return themeConfig[key];
}
```

Is that right?

Reply:
yes something like that, though I do wonder where this `themeConfig` come from :P

Reply:
In my mind, the "themeConfig" here is context aware (so back to the ThemeProvider just setting a context value)

Reply:
🤔 I'll need to think about this.

My initial thought is just that it would pluck it from the passed in `config` but that leaves out the different "modes". Not sure what to do about that as the previous way modes were handled was to just swap out the variables and the CSS would update automatically, but if we're returning raw values from `get` that's not going to be possible.

Likewise, `get` needs to be usable outside the context of React (like in a `styles.js` file where we're just calling `css` and `styled`) so we cannot rely on all those values being inside of a React context (plus adding React context everywhere adds a performance hit that I'm pretty sure Q spent a good deal of time measuring).

Reply:
> I'd love other thoughts on this as well. This is something that can have some impact long term (WP user profiles, theming components) so yeah more thoughts and opinions here would be good.

I lean towards playing nicely with existing systems and starting small. As Riad points out, the impact of these changes is unclear, and there's the risk of starting _de facto_ APIs that will later cost us to support.

Reply:
What if we change to prefix to `--wp-experimental` for all these variables, leave the rest the way it is, and then re-evaluate how to do this once we've integrated the `styles` package which actually contains the theme? It might be more helpful to brainstorm solutions at that point than now when we only have half the context.

Reply:
I would defer to Riad or Miguel, whose concerns I share: we should move carefully and only introduce components, systems, and custom properties when we know for a fact that they will be needed and used. 

To use an example: color custom properties. I've seen a tendency to design big palettes and adding them all as custom properties to be conveniently available. In practice, they end up being dead weight on most pages, and causing developer confusion with the breadth of choice (maybe gray-200 was chosen when gray-300 should've been used).

That is to say: these properties are only ever useful if they will be used, and ideally used often enough to deserve such stature. I do recognize that it's a spectrum ranging from overwhelming palettes to lack of flexibility, and that the ideal spot, probably, is somewhere in the middle. But it is an argument for moving slowly and only when we have some confidence in the system.

Reply:
[@wordpress/base-styles](https://github.com/WordPress/gutenberg/tree/trunk/packages/base-styles) contain several files and a huge amount of code. Ensuring it stays in sync with CSS-in-JS bits might be a huge challenge unless we find a good way to share them. 

> By this do you mean that the new package would just create the theme variables into the :root scope?

This sounds like a reasonable proposal because we could use those variables in other places – outside of `@wordpress/components`.

Reply:
> What if we change to prefix to --wp-experimental for all these variables, leave the rest the way it is, and then re-evaluate how to do this once we've integrated the styles package which actually contains the theme? It might be more helpful to brainstorm solutions at that point than now when we only have half the context.

That's fine I guess but it's definitely something we need to address.